### ISSUE TITLE
Login time increases strongly if more than one domain is configured

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#847]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/843
- Created at 2011-04-08 15:33:05 by [sbose](https://pagure.io/user/sbose)
- Closed at 2017-03-30 13:40:46 as Fixed
- Assigned to [fidencio](https://pagure.io/user/fidencio)
---

If there are multiple domains configured, e.g. 'domains=d1,d2,d3', the time until a user form domain d3 is logged in increases (2-3 times) compared to a configuration with only one domain 'domains=d3'.

### Comments

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2011-04-08 21:46:07](https://pagure.io/SSSD/sssd/issue/843#232487)**

https://bugzilla.redhat.com/show_bug.cgi?id=694905

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2011-04-14 14:58:32](https://pagure.io/SSSD/sssd/issue/843#232488)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.7.0
patch:  => 0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2011-07-22 22:00:44](https://pagure.io/SSSD/sssd/issue/843#232489)**

Pushing out as multiple domain is not a popular use case yet.

milestone: SSSD 1.8.0 => SSSD 1.9.0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-01-03 15:20:09](https://pagure.io/SSSD/sssd/issue/843#232490)**

(In #1126) There are a number of improvements we should work on while fixing this. In particular, we should seek to resolve the performance issue identified in bug #843.

The current approach behaves as follows (I will use {{{getpwnam()}}} as the representative function, though all are equivalent).

The NSS responder will check the cache for the first domain in the list. If it is empty or expired, the Responder will query the Data Provider. If the Data Provider returns DP_ERR_OK, the cache is checked again. If the entry is still not present, it advances to the next domain. This pattern repeats.

The problem with this approach is that it means that for a configuration of N domains, if a user is found in the last domain in the list, SSSD will go to the server N-1 times before ever checking the cache of the correct domain.

The proposed new approach is for us to search each domain's cache first. If any domain has the requested, non-expired entry, it should be returned immediately. If all domains lack the entry, only then should we begin to iterate through them. For NSS lookups, we should NOT limit ourselves only to the domain with an expired entry for the user, because we need to be able to catch new users being added to earlier domains. So in the complete cache-miss case, we will still iterate through all domains, but a cache-hit will not result in network activity.

For PAM, we will take an additional shortcut. Since it is reasonable to assume that the login program will have made a {{{getpwnam()}}} call prior to starting the PAM conversation, we will limit our mandatory {{{initgroups()}}} lookup to only domains with an expired entry (if one exists, otherwise we will try them in series).

blockedby:  => 1126

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-01-16 16:35:36](https://pagure.io/SSSD/sssd/issue/843#232491)**

Fields changed

blockedby:  => 
blocking:  => 
milestone: SSSD 1.9.0 => SSSD Infopipe
rhbz:  =>

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-01-16 17:13:49](https://pagure.io/SSSD/sssd/issue/843#232492)**

Fields changed

owner: somebody => nalin

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-01-19 03:13:31](https://pagure.io/SSSD/sssd/issue/843#232493)**

Fields changed

rhbz:  => 0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-02-02 21:11:43](https://pagure.io/SSSD/sssd/issue/843#232494)**

Fields changed

feature_milestone:  =>

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-02-02 21:12:10](https://pagure.io/SSSD/sssd/issue/843#232495)**

Fields changed

milestone: SSSD Infopipe Feature => NEEDS_TRIAGE

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-02-05 20:40:05](https://pagure.io/SSSD/sssd/issue/843#232496)**

We also need to look into how the new responders such as sudo, ssh and autofs implement the loop and convert them into using the same logic as services use.

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-02-09 16:06:14](https://pagure.io/SSSD/sssd/issue/843#232497)**

Fields changed

component: PAM => NSS
milestone: NEEDS_TRIAGE => SSSD 1.9.0
owner: nalin => sgallagh

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-07-12 15:39:35](https://pagure.io/SSSD/sssd/issue/843#232498)**

Fields changed

milestone: SSSD 1.9.0 => SSSD 1.11.0 (LTM)
type: enhancement => defect

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-08-16 19:15:14](https://pagure.io/SSSD/sssd/issue/843#232499)**

Fields changed

proposed_priority:  => Nice to have

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-08-16 19:33:47](https://pagure.io/SSSD/sssd/issue/843#232500)**

Fields changed

blockedby:  => 1126

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-09-04 23:20:30](https://pagure.io/SSSD/sssd/issue/843#232501)**

Moving all the features planned for 1.10 release into 1.10 beta.

milestone: SSSD 1.11.0 (LTM) => SSSD 1.10 beta

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-09-04 23:50:12](https://pagure.io/SSSD/sssd/issue/843#232502)**

Fields changed

priority: major => minor

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2012-12-20 23:33:42](https://pagure.io/SSSD/sssd/issue/843#232503)**

Fields changed

selected:  => Not need

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2013-01-02 15:32:08](https://pagure.io/SSSD/sssd/issue/843#232504)**

Moving tickets that are not a priority for SSSD 1.10 into the next release.

milestone: SSSD 1.10 beta => SSSD 1.11 beta

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2014-02-06 17:08:39](https://pagure.io/SSSD/sssd/issue/843#232505)**

Putting up to NEEDS_TRIAGE so we can discuss the priority on the next meeting.

changelog:  => 
design:  => 
design_review:  => 0
fedora_test_page:  => 
keywords:  => localaccounts
review:  => 0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2014-02-06 17:08:49](https://pagure.io/SSSD/sssd/issue/843#232506)**

Fields changed

milestone: SSSD 1.13 beta => NEEDS_TRIAGE

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2014-02-13 15:54:11](https://pagure.io/SSSD/sssd/issue/843#232507)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.13 beta

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2014-09-26 18:49:56](https://pagure.io/SSSD/sssd/issue/843#232508)**

Fields changed

mark:  => 0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2015-02-10 16:51:20](https://pagure.io/SSSD/sssd/issue/843#232509)**

Required for local account support.

milestone: SSSD 1.13 beta => SSSD 1.13 backlog

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2015-02-12 20:14:53](https://pagure.io/SSSD/sssd/issue/843#232510)**

Fields changed

milestone: SSSD 1.13 backlog => SSSD 1.14 beta
priority: minor => major

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2015-07-28 20:12:00](https://pagure.io/SSSD/sssd/issue/843#232511)**

Fields changed

priority: major => critical
sensitive:  => 0

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2016-02-16 11:43:54](https://pagure.io/SSSD/sssd/issue/843#232512)**

Pavel already started submitting patches that use the new cache_req code but until we switch the responders to the new code, we can't make this switch either, sorry.

Therefore, I'm moving this ticket out of the 1.14 bucket for capacity reasons.

milestone: SSSD 1.14 beta => SSSD 1.15 beta

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-02-06 15:55:38](https://pagure.io/SSSD/sssd/issue/843#232513)**

This would be solved together with using shortname for trusted domains, so Fabiano is probably the right owner for both.

milestone: SSSD Future releases (no date set yet) => SSSD 1.15.2
owner: sgallagh => fidencio

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-02-24 14:45:50](https://pagure.io/SSSD/sssd/issue/843#232514)**

**Metadata Update from @sbose**:
- Issue assigned to fidencio
- Issue marked as blocked by: #2228
- Issue marked as depending on: #1126
- Issue set to the milestone: SSSD 1.15.2

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-15 10:23:07](https://pagure.io/SSSD/sssd/issue/843#431412)**

This issue was fixed in the following commits:
 * 606015a
 * 8bb6680
 * 9286d0d
 * 7cd2264
 * 828fe75

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-15 10:23:15](https://pagure.io/SSSD/sssd/issue/843#431413)**

This issue was fixed in the following commits:
 * 606015a
 * 8bb6680
 * 9286d0d
 * 7cd2264
 * 828fe75

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-15 10:24:34](https://pagure.io/SSSD/sssd/issue/843#431416)**

but I can't close the ticket because of the blocks/depends on which is a pagure bug..

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-15 10:25:54](https://pagure.io/SSSD/sssd/issue/843#431419)**

but I can't close the ticket because of the blocks/depends on which is a pagure bug..

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-30 13:40:35](https://pagure.io/SSSD/sssd/issue/843#434216)**

**Metadata Update from @jhrozek**:
- Custom field design_review reset (from 0)
- Custom field mark reset (from 0)
- Custom field patch reset (from 0)
- Custom field review reset (from 0)
- Custom field sensitive reset (from 0)
- Custom field testsupdated reset (from 0)
- Issue **un**marked as depending on: #2228

---

**Comment from [sbose](https://pagure.io/user/sbose) at [2017-03-30 13:40:49](https://pagure.io/SSSD/sssd/issue/843#434218)**

**Metadata Update from @jhrozek**:
- Custom field design_review reset (from false)
- Custom field mark reset (from false)
- Custom field patch reset (from false)
- Custom field review reset (from false)
- Custom field sensitive reset (from false)
- Custom field testsupdated reset (from false)
- Issue close_status updated to: Fixed
- Issue status updated to: Closed (was: Open)
